lib/logger/lib/logger.ex

```diff
@@ -463,41 +441,39 @@ defmodule Logger do
   """
   @spec metadata(metadata) :: :ok
   def metadata(keyword) do
-    {enabled?, metadata} = __metadata__()
-    Process.put(@metadata, {enabled?, into_metadata(keyword, metadata)})
-    :ok
-  end
+    case :logger.get_process_metadata() do
+      :undefined ->
+        :ok = :logger.set_process_metadata(filter_out_nils(keyword))
 
-  defp into_metadata([], metadata), do: metadata
-  defp into_metadata(keyword, metadata), do: into_metadata(keyword, [], metadata)
+      map when is_map(map) ->
+        merged = Enum.into(keyword, map)
+        metadata = filter_out_nils(merged)
 
-  defp into_metadata([{key, nil} | keyword], prepend, metadata) do
-    into_metadata(keyword, prepend, :lists.keydelete(key, 1, metadata))
-  end
-
-  defp into_metadata([{key, _} = pair | keyword], prepend, metadata) do
-    into_metadata(keyword, [pair | prepend], :lists.keydelete(key, 1, metadata))
+        :ok = :logger.set_process_metadata(metadata)
+    end
   end
 
-  defp into_metadata([], prepend, metadata) do
-    prepend ++ metadata
+  defp filter_out_nils(enumerable) do
+    for {_k, v} = elem <- enumerable, v != nil, into: %{}, do: elem
   end
 
   @doc """
   Reads the current process metadata.
   """
   @spec metadata() :: metadata
   def metadata() do
-    __metadata__() |> elem(1)
+    case :logger.get_process_metadata() do
+      :undefined -> []
+      map when is_map(map) -> Map.to_list(map)
+    end
   end
 
   @doc """
   Resets the current process metadata to the given keyword list.
   """
   @spec reset_metadata(metadata) :: :ok
   def reset_metadata(keywords \\ []) do
-    {enabled?, _metadata} = __metadata__()
-    Process.put(@metadata, {enabled?, []})
+    :ok = :logger.set_process_metadata(%{})
     metadata(keywords)
```

Comment:
This function could be more efficiently implemented as:

    :ok = :logger.set_process_metadata(filter_out_nils(keyword))